Expand WordPress Intelligence support to other locales

kean opened the pull request:
## What

Fixes [CMM-762: Excerpts are created in the system language not the content language](https://linear.app/a8c/issue/CMM-762/excerpts-are-created-in-the-system-language-not-the-content-language) and [CMM-798: Add support for other locales](https://linear.app/a8c/issue/CMM-798/add-support-for-other-locales)

I smoke-tested it by generating a post in Spanish and testing that the excerpts are also in Spanish. It should work for any other scenarios as well. 

<img width="320" alt="Screenshot 2025-11-28 at 4 10 02 PM" src="https://github.com/user-attachments/assets/5166ca44-6f34-4c12-ab43-a4734f147c1c" />

## How

- Move the Intelligence-related code to a new `WordPressModule` and use it as a namespace (rename some of the features)
- Add an extensive lists of tests primarily focusing on locales. The tests are non-deterministic and slow, so they are not intended to be run on every change (for now). They are designed for semi-manually verifying the changes to prompts and other parts of `WordPressIntelligence`.

It's a large PR, but the majority of the changes are in unit tests. The main changes to look for is in prompts.

kean commented on WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift:
```diff
@@ -80,8 +81,10 @@ public enum FeatureFlag: Int, CaseIterable {
         case .intelligence:
-            let languageCode = Locale.current.language.languageCode?.identifier
-            return (languageCode ?? "en").hasPrefix("en")
+            guard #available(iOS 26, *) else {
```
It's technically redundant; just some more defensive code.

crazytonyli commented on Modules/Sources/WordPressShared/Intelligence/IntelligenceService.swift:
```diff
@@ -124,10 +128,11 @@ public actor IntelligenceService {
         with fewer than 10 words.
 
         The summary should be clear, informative, and written in a neutral tone.
+        You MUST generate the summary in the same language as the support request.
```
Nitpick: I think the prompt can be more direct here: "You MUST generate the summary in the `(locale.identifier)` locale.", and remove the `makeLocaleInstructions()` above?

crazytonyli commented:
The post summary generation is not updated in this PR. Probably because that went to a separate code path: `LanguageModelHelper.makeGenerateExcerptPrompt`.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>.</p>

<p>If the change includes adding, removing, or editing a dependency please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</p>

<p>If the change to the <code>Package.swift</code> did not modify dependencies, ignoring this warning should be safe, but we recommend double checking and running the package resolution just in case.<br />
.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean commented:
I added a significant amount of automated tests, improved the prompts to get it to produce the correct language output at a much higher rate, and updated the PR description. 

I also have another upcoming PR with additional LLM-based evaluation for automation.

kean commented:
There are a couple of SwiftLint issues which I will be addressing before merge.

kean committed:
Add support for more locales

kean commented on Modules/Tests/WordPressIntelligenceTests/Helpers/ExcerptTestOutput.swift:
```diff
@@ -0,0 +1,63 @@
+    /// This output can be extracted and evaluated by external tools.
+    private func record() throws {
+        // Encode to JSON
+        let encoder = JSONEncoder()
```
This is required for https://github.com/wordpress-mobile/WordPress-iOS/pull/25059. It's a small part I included in this PR for simplicity of managing commits.

crazytonyli commented on Modules/Sources/WordPressIntelligence/UseCases/TagSuggestionGenerator.swift:
```diff
@@ -0,0 +1,130 @@
+    ///   - siteTags: Existing tags from the site
+    ///   - postTags: Tags already added to this post
+    /// - Returns: Formatted prompt string ready for the language model
+    public func makePrompt(post: String, siteTags: [String], postTags: [String]) async -> String {
```
Is this function `async` because `detectLanguage` can be slow?

kean replied:
Yes, it to make `detectLanguage` and everything else run in the background (it does).

crazytonyli commented on Modules/Sources/WordPressIntelligence/UseCases/SupportTicketSummaryGenerator.swift:
```diff
@@ -0,0 +1,44 @@
+            with fewer than 10 words.
+
+            The summary should be clear, informative, and written in a neutral tone.
+            You MUST generate the summary in the same language as the support request.
```
Does this prompt work with giving LLM the `detectLanguage` result?

crazytonyli replied:
BTW, I think it's okay to generate a summary in the device language in this context. The summary is for the customer, not website visitors.

kean replied:
> in the device language in this context. The summary is for the customer, not website visitors.

I was just thinking about it – do we want to translate the post or not and whether it should be the same thing as "summarize". I would argue it should be two separate features.

I do plan to add translation, starting with comments [CMM-744: On-device translation for comments](https://linear.app/a8c/issue/CMM-744/on-device-translation-for-comments)

crazytonyli commented on Modules/Sources/WordPressIntelligence/UseCases/PostSummaryGenerator.swift:
```diff
@@ -0,0 +1,94 @@
+    /// - Returns: Formatted prompt string ready for the language model
+    public func makePrompt(content: String) async -> String {
+        let extractedContent = IntelligenceService.extractRelevantText(from: content, ratio: 0.8)
+        let language = IntelligenceService.detectLanguage(from: extractedContent)
```
We probably want the generated summary to be in the device language. The summary is for the current user after all.

I guess the AI summary feature can be used as a quick way to translate and summarize if I don't understand the original post's language.

kean commented:
> The excerpt works for me now. But the suggested tags are still unrelated to the post content.

The current prompt prioritizes language of the existing tags for generation. Are the tags from the screenshot on your site? These are taken into account when generating suggestions. Do they have any relation with the content?

I suggest to file it as a bug with input/output. I'll include it in the testing suite later. The unit tests seem to be pretty high confidence that the generation happens in the expected language. I can't guarantee it will work for every use case with these models. I plan to use the LLM-as-a-judge evaluation to improve it further. I think it looks good enough for current iteration based on the unit tests.

kean commented:
> The excerpt works for me now. But the suggested tags are still unrelated to the post content.

I did a bit more testing, and I confirm this is a potential issues, but it's not directly related to localization. I opened [CMM-1073: Improve quality of suggested tags](https://linear.app/a8c/issue/CMM-1073/improve-quality-of-suggested-tags) to track it. The issue is that the prompt prioritizes current site tags, their language, and format – which is generally a good thing for real sites. If you have a test site with a bunch of tags unrelated to the posts, it will not work as well, which is also expected. However, I think it still needs improvement as it should suggest tags only if there is a high level of confidence they match the text. Having said that, I'm not completely sure because you may be tagging based on some other criteria, so showing _some_ suggestions can be better than showing none. I think the app might need do a little of work in the background to try and find relationships between your posts and tags before making suggestions.

> Also, maybe we don't need to give the option to generate excerpts if the content is super short? It'd be pretty tricky for LLM to generate accurate excerpts if the post content is short.

It is an existing ticket [CMM-763: If the post content is too short, the results are not helpful](https://linear.app/a8c/issue/CMM-763/if-the-post-content-is-too-short-the-results-are-not-helpful). I don't know how to define "short", so I haven't looked into it yet. It doesn't seem like an issue that should be addressed as nothing bad happens if you do it and it's an unlikely scenario a user would use generation in the first place. For tags, I'll look into it in the scope of the previous issue I linked.
